modules/cudf/src/dataframe/print.ts

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) 2021, NVIDIA CORPORATION.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import {DataFrame} from '../data_frame';
+import {Series} from '../series';
+import {Int32, Utf8String} from '../types/dtypes';
+import {TypeMap} from '../types/mappings';
+
+export type DisplayOptions = {
+  /**
+   * The maximum number of columns to display before truncation. Truncation may also occur
+   * if the total printed width in characters exceeds `widht
+   *
+   */
+  maxColumns?: number,
+
+  /**
+   * The maximum width in characters of a column when printing a DataFram. When the column *
+   * overflows, a "..." placeholder is embedded in the output. A value of zero means unlimited.
+   * (default: 50)
+   */
+  maxColWidth?: number,
+
+  /**
+   * The maximum number of rows to output when printing a DataFrame.  A value of zero means
+   * unlimited. (default: 60)
+   */
+  maxRows?: number,
+
+  /**
+   * Width of the display in characters. (default: 80)
+   */
+  width?: number,
+
+}
+
+export class DataFrameFormatter<T extends TypeMap> {
+  private obj: DataFrame;
+  private maxColumns: number;
+  private maxColWidth: number;
+  private maxRows: number;
+  private width: number;
+  private htrunc: boolean;
+  private vtrunc: boolean;
+
+  constructor(options: DisplayOptions, obj: DataFrame<T>) {
+    this.maxColumns  = options.maxColumns ?? 20;
+    this.maxColWidth = options.maxColWidth ?? 50;
+    this.maxRows     = options.maxRows ?? 60;
+    this.width       = options.width ?? 80;
+    this.htrunc      = false;
+    this.vtrunc      = false;
+
+    let tmp = this.preprocess(obj);
+    while (this._totalWidth(tmp) > this.width) {
+      this.htrunc = true;
+      const names = [...tmp.names];
+      const N     = Math.ceil((names.length - 1) / 2);
+      tmp         = tmp.drop([names[N]]);
+    }
```

Comment:
This loop could get in to trouble. I see two options:

* bail if new total width ever equals prev total width
* set a min total width and error earlier

Reply:
Yeah setting a min value for `width` seems reasonable to me.

Reply:
I will probably add a min `width` check (5 characters?) but there are lots of interactions so I think I will add the clunky `current = prev` check to. But then there's a question of what to do:

* if things get down to zero column / can't get total width < with

Is that an *error*? Or is that "return an empty string"?

Reply:
I'd personally return an empty string, however, is there some convention that the `toString()` methods follow in javascript?

Reply:
I think I'm inclined that way as well. @trxcllnt do you have a preference here?

Reply:
We could truncate the last column values if the user requests a pathologically small max width... Maybe our minimum should be 4 -- 1 character of data + an ellipses?

Reply:
That's an interesting idea. That is not what pandas does but I can look into it if

Reply:
If our going-in position is default to infinite width, then TBH this seems like a fair amount of complexity for a really rare/unimportant case. I did set a min col width of 4, so there will always be at least some output (even if its just a column of elipses). Any objection if I leave it at that?

Reply:
If I'm reading this right, we'll drop all the columns if the max width is smaller than the width of the smallest column? I can live with that.